gqcp/tests/QCMethod/CC/CCD0_test.cpp

```diff
@@ -0,0 +1,257 @@
+#define BOOST_TEST_MODULE "CCD0"
+
+#include <boost/test/unit_test.hpp>
+#include <iostream>
+#include <fstream>
+#include <string>
+
+#include "QCMethod/CC/CCD0.hpp"
+#include "Basis/ScalarBasis/GTOBasisSet.hpp"
+#include "Basis/ScalarBasis/GTOShell.hpp"
+#include "Basis/SpinorBasis/RSpinorBasis.hpp"
+#include "Basis/SpinorBasis/USpinorBasis.hpp"
+#include "Molecule/Molecule.hpp"
+#include "Operator/SecondQuantized/SQTwoElectronOperator.hpp"
+#include "Operator/SecondQuantized/SQHamiltonian.hpp"
+#include "Mathematical/Representation/QCRankFourTensor.hpp"
+#include "Operator/FirstQuantized/NuclearRepulsionOperator.hpp"
+#include "Operator/FirstQuantized/OverlapOperator.hpp"
+#include "Operator/SecondQuantized/SQOneElectronOperator.hpp"
+#include "Operator/SecondQuantized/SQTwoElectronOperator.hpp"
+#include "QCMethod/HF/RHFSCFEnvironment.hpp"
+#include "QCMethod/HF/RHFSCFSolver.hpp"
+#include "QCMethod/HF/DiagonalRHFFockMatrixObjective.hpp"
+#include "QCMethod/HF/RHF.hpp"
+#include "QCMethod/QCStructure.hpp"
+#include "Mathematical/Representation/BlockRankFourTensor.hpp"
+
+GQCP::QCRankFourTensor<double> twoElectronAntiSymmetrized(const GQCP::QCRankFourTensor<double>& gs) {
+    size_t dim = gs.dimension();
+    GQCP::QCRankFourTensor<double> ls(dim);
+
+    for (size_t p=0 ; p!=dim ; p++){
+        for (size_t q=0 ; q!=dim ; q++){
+            for (size_t r=0 ; r!=dim ; r++){
+                for (size_t s=0 ; s!=dim ; s++){
+                    ls(p, q, r, s) = 2*gs(p, q, r, s) - gs(p, s, r, q);
+                }
+            }
+        }
+    }
+    return ls;
+}
+
+double calculateProjection(const GQCP::SQHamiltonian<double>& sq_hamiltonian, GQCP::BlockRankFourTensor<double>& ts, const GQCP::ScalarSQOneElectronOperator<double>& f,
+                        const size_t a, const size_t b, const size_t i, const size_t j){
+    const auto hs = sq_hamiltonian.core().parameters(0);
```

Comment:
Since the parameters themselves exist already in memory, it’s better to add a ‘reference’ here: ‘const auto& h = sq_hamiltonian.core().parameters()’.

Furthermore, it seems that your amplitudes are not modified: could you add a ‘const’ to indicate this?

Reply:
Foolish mistake indeed, copying large tensors could be costly 😁.